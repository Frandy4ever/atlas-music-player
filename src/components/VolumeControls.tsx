import { useState } from "react";
import { Volume2, VolumeOff } from "lucide-react";

interface VolumeControlsProps {
  onVolumeChange: (volume: number) => void;
  onLyricsToggle: () => void;
}

const VolumeControls: React.FC<VolumeControlsProps> = ({ onVolumeChange, onLyricsToggle }) => {
  const [volume, setVolume] = useState(50);
  const [showPercentage, setShowPercentage] = useState(false);
  const [showTooltip, setShowTooltip] = useState(false);

  const isMuted = volume === 0;

  // Determine icon color based on state
  const getIconColor = (active: boolean) =>
    active
      ? "text-blue-600 dark:text-red-600 hover:text-blue-400 dark:hover:text-red-300"
      : "text-gray-800 dark:text-white hover:text-blue-400 dark:hover:text-red-400";

  const handleVolumeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newVolume = Number(event.target.value);
    setVolume(newVolume);
    onVolumeChange(newVolume);
  };

  const getVolumeColor = () => {
    if (volume <= 30) return "bg-black text-white accent-black";
    if (volume > 30 && volume <= 50) return "bg-stone-950 text-white accent-stone-950";
    if (volume > 50 && volume <= 80) return "bg-blue-950 dark:bg-red-950 text-white accent-blue-950 dark:accent-red-950";
    return "bg-blue-600 dark:bg-red-600 text-black accent-blue-600 dark:accent-red-600";
  };

  return (
    <div className="flex items-center gap-3 w-full">
      {/* Volume Icon */}
      <button className={getIconColor(isMuted)} onClick={() => setVolume(volume > 0 ? 0 : 50)}>
        {isMuted ? <VolumeOff size={24} /> : <Volume2 size={24} />}
      </button>

      {/* Volume Slider */}
      <div
        className="relative flex-grow"
        onMouseEnter={() => setShowPercentage(true)}
        onMouseLeave={() => setShowPercentage(false)}
      >
        <div
          className={`absolute -top-3 left-1/2 -translate-x-1/2 text-sm font-semibold px-1 rounded-md transition-all ${showPercentage ? "opacity-100" : "opacity-0"
            } ${getVolumeColor()}`}
          style={{ left: `${volume}%`, transform: "translate(-50%, 0)" }}
        >
          {volume}%
        </div>

        <input
          type="range"
          min="0"
          max="100"
          value={volume}
          onChange={handleVolumeChange}
          onMouseDown={() => setShowPercentage(true)}
          onMouseUp={() => setShowPercentage(false)}
          className={`h-1 rounded-lg cursor-pointer transition-all ${getVolumeColor()} w-full`}
        />
      </div>

      {/* Lyrics Icon */}
      <div className="relative">
        <button
          className="text-gray-800 dark:text-white hover:text-blue-400 dark:hover:text-red-400"
          onClick={onLyricsToggle}
          onMouseEnter={() => setShowTooltip(true)}
          onMouseLeave={() => setShowTooltip(false)}
        >
          <span className="material-symbols-outlined text-2xl">lyrics</span>
        </button>

        {showTooltip && (
          <div className="absolute -top-5 left-1/2 -translate-x-1/2 bg-black text-white text-xs p-1 rounded shadow-md">
            Lyrics
          </div>
        )}
      </div>
    </div>
  );

};

export default VolumeControls;
