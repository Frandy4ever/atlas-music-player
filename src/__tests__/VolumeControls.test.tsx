import { render } from "@testing-library/react";
import VolumeControls from "../components/VolumeControls";

describe("VolumeControls Component", () => {
  test("matches snapshot with default volume", () => {
    const { asFragment } = render(
      <VolumeControls onVolumeChange={() => {}} onLyricsToggle={() => {}} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
