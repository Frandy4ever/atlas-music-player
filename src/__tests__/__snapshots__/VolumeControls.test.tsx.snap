// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VolumeControls Component > matches snapshot with default volume 1`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-3 w-full"
  >
    <button
      class="text-gray-800 dark:text-white hover:text-blue-400 dark:hover:text-red-400"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-volume2 lucide-volume-2"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
        />
        <path
          d="M16 9a5 5 0 0 1 0 6"
        />
        <path
          d="M19.364 18.364a9 9 0 0 0 0-12.728"
        />
      </svg>
    </button>
    <div
      class="relative flex-grow"
    >
      <div
        class="absolute -top-3 left-1/2 -translate-x-1/2 text-sm font-semibold px-1 rounded-md transition-all opacity-0 bg-stone-950 text-white accent-stone-950"
        style="left: 50%; transform: translate(-50%, 0);"
      >
        50%
      </div>
      <input
        class="h-1 rounded-lg cursor-pointer transition-all bg-stone-950 text-white accent-stone-950 w-full"
        max="100"
        min="0"
        type="range"
        value="50"
      />
    </div>
    <div
      class="relative"
    >
      <button
        class="text-gray-800 dark:text-white hover:text-blue-400 dark:hover:text-red-400"
      >
        <span
          class="material-symbols-outlined text-2xl"
        >
          lyrics
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
